lwgenerate/hcl.go

```diff
@@ -0,0 +1,366 @@
+	switch v := val.(type) {
+	case string, int, bool:
+		value, err := convertTypeToCty(v)
+		if err != nil {
+			return err
+		}
+		block.Body().SetAttributeValue(key, value)
+	case hcl.Traversal:
+		block.Body().SetAttributeTraversal(key, v)
+	case map[string]interface{}:
+		data := map[string]cty.Value{}
+		for attrKey, attrVal := range v {
+			value, err := convertTypeToCty(attrVal)
+			if err != nil {
+				return err
+			}
+			data[attrKey] = value
+		}
+		block.Body().SetAttributeValue(key, cty.ObjectVal(data))
+	default:
+		return errors.New("Unknown type")
+	}
+
+	return nil
+}
+
+// Helper to create various types of new hclwrite.Block using generic inputs
+func HclCreateGenericBlock(hcltype string, labels []string, attr map[string]interface{}) (*hclwrite.Block, error) {
+	block := hclwrite.NewBlock(hcltype, labels)
+
+	// Source and version require some special handling, should go at the top of a block declaration
+	sourceFound := false
+	versionFound := false
+
+	// We need/want to guarantee the ordering of the attributes, do that here
+	var keys []string
+	for k := range attr {
+		switch k {
+		case "source":
+			sourceFound = true
+		case "version":
+			versionFound = true
+		default:
+			keys = append(keys, k)
+		}
+	}
+	sort.Strings(keys)
+
+	if sourceFound || versionFound {
+		var newKeys []string
+		if sourceFound {
+			newKeys = append(newKeys, "source")
+		}
+		if versionFound {
+			newKeys = append(newKeys, "version")
+		}
+		keys = append(newKeys, keys...)
+	}
+
+	// Write block data
+	for _, key := range keys {
+		val := attr[key]
+		if err := setBlockAttributeValue(block, key, val); err != nil {
+			return nil, err
+		}
+	}
+
+	return block, nil
+}
+
+// Create tokens for map of traversals.  Used as a workaround for writing complex types where the built-in
+// SetAttributeValue won't work
+func createMapTraversalTokens(input map[string]string) hclwrite.Tokens {
+	// Sort input
+	keys := []string{}
+	for k := range input {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	tokens := hclwrite.Tokens{
+		{Type: hclsyntax.TokenOBrace, Bytes: []byte("{"), SpacesBefore: 1},
+		{Type: hclsyntax.TokenNewline, Bytes: []byte("\n")},
+	}
+
+	for _, k := range keys {
+		tokens = append(tokens, []*hclwrite.Token{
+			{Type: hclsyntax.TokenStringLit, Bytes: []byte(k)},
+			{Type: hclsyntax.TokenEqual, Bytes: []byte("=")},
+			{Type: hclsyntax.TokenStringLit, Bytes: []byte(" " + input[k]), SpacesBefore: 1},
+			{Type: hclsyntax.TokenNewline, Bytes: []byte("\n")},
+		}...)
+	}
+
+	tokens = append(tokens, []*hclwrite.Token{
+		{Type: hclsyntax.TokenNewline},
+		{Type: hclsyntax.TokenCBrace, Bytes: []byte("}")},
+	}...)
+
+	return tokens
+}
+
+// Convert blocks to a string
+func CreateHclStringOutput(blocks []*hclwrite.Block) string {
+	file := hclwrite.NewEmptyFile()
+	body := file.Body()
+
+	for _, b := range blocks {
+		body.AppendBlock(b)
+		body.AppendNewline()
+	}
+	return string(file.Bytes())
```

Comment:
For some reason there are trailing spaces and carrier returns at the end, this trims them all
```suggestion
	return strings.TrimRight(string(file.Bytes()), "\n")
```

Reply:
👍

Reply:
Two things were causing this 

1) I added new lines to each block to provide proper spacing (L297 above)
2) The examples were using `fmt.Println` which added yet another space (HCL examples are printing single blocks so each block would have this behavior giving double spacing vs AWS which is one slice of blocks so it only had 1 trailing space)

I've resolved both of these by fixing the examples and by only appending newlines to non-last blocks.

Reply:
👌🏽